File: server/storage.ts
import { 
  users, menuItems, orders, contactMessages,
  type User, type InsertUser,
  type MenuItem, type InsertMenuItem,
  type Order, type InsertOrder,
  type ContactMessage, type InsertContactMessage
} from "@shared/schema";

export interface IStorage {
  // Users
  getUser(id: number): Promise<User | undefined>;
  getUserByUsername(username: string): Promise<User | undefined>;
  createUser(user: InsertUser): Promise<User>;
  
  // Menu Items
  getAllMenuItems(): Promise<MenuItem[]>;
  getMenuItemsByCategory(category: string): Promise<MenuItem[]>;
  createMenuItem(item: InsertMenuItem): Promise<MenuItem>;
  
  // Orders
  createOrder(order: InsertOrder): Promise<Order>;
  getOrder(id: number): Promise<Order | undefined>;
  getAllOrders(): Promise<Order[]>;
  updateOrderStatus(id: number, status: string): Promise<Order | undefined>;
  
  // Contact Messages
  createContactMessage(message: InsertContactMessage): Promise<ContactMessage>;
  getAllContactMessages(): Promise<ContactMessage[]>;
}

export class MemStorage implements IStorage {
  private users: Map<number, User>;
  private menuItems: Map<number, MenuItem>;
  private orders: Map<number, Order>;
  private contactMessages: Map<number, ContactMessage>;
  private currentUserId: number;
  private currentMenuItemId: number;
  private currentOrderId: number;
  private currentContactMessageId: number;

  constructor() {
    this.users = new Map();
    this.menuItems = new Map();
    this.orders = new Map();
    this.contactMessages = new Map();
    this.currentUserId = 1;
    this.currentMenuItemId = 1;
    this.currentOrderId = 1;
    this.currentContactMessageId = 1;
    
    // Initialize with menu items
    this.initializeMenuItems();
  }

  private initializeMenuItems() {
    const menuData: InsertMenuItem[] = [
      // Starters
      { name: "Samosas (2 pieces)", description: "Crispy pastries filled with spiced potatoes and peas", price: "6.99", category: "Starters", available: true },
      { name: "Pakoras", description: "Mixed vegetable fritters with mint chutney", price: "8.99", category: "Starters", available: true },
      { name: "Aloo Tikki", description: "Pan-fried potato patties with tamarind sauce", price: "7.99", category: "Starters", available: true },
      { name: "Chana Chaat", description: "Spiced chickpea salad with yogurt and chutneys", price: "9.99", category: "Starters", available: true },
      
      // Main Courses
      { name: "Butter Chicken", description: "Tender chicken in rich tomato and cream sauce", price: "16.99", category: "Main Courses", available: true },
      { name: "Dal Makhani", description: "Slow-cooked black lentils with butter and cream", price: "14.99", category: "Main Courses", available: true },
      { name: "Palak Paneer", description: "Fresh spinach curry with cottage cheese", price: "15.99", category: "Main Courses", available: true },
      { name: "Chicken Biryani", description: "Aromatic basmati rice with spices and chicken", price: "18.99", category: "Main Courses", available: true },
      { name: "Vegetable Biryani", description: "Aromatic basmati rice with spices and vegetables", price: "16.99", category: "Main Courses", available: true },
      { name: "Rajma", description: "Kidney beans in spiced tomato gravy", price: "13.99", category: "Main Courses", available: true },
      { name: "Chole Bhature", description: "Spiced chickpeas with fluffy fried bread", price: "15.99", category: "Main Courses", available: true },
      
      // Desserts
      { name: "Gulab Jamun (2 pieces)", description: "Soft milk dumplings in rose-scented syrup", price: "5.99", category: "Desserts", available: true },
      { name: "Kheer", description: "Traditional rice pudding with cardamom and nuts", price: "6.99", category: "Desserts", available: true },
      { name: "Kulfi", description: "Dense, creamy Indian ice cream with pistachios", price: "4.99", category: "Desserts", available: true },
      { name: "Rasmalai", description: "Soft cheese patties in sweetened milk", price: "7.99", category: "Desserts", available: true },
    ];

    menuData.forEach(item => {
      const id = this.currentMenuItemId++;
      const menuItem: MenuItem = { ...item, id };
      this.menuItems.set(id, menuItem);
    });
  }

  // Users
  async getUser(id: number): Promise<User | undefined> {
    return this.users.get(id);
  }

  async getUserByUsername(username: string): Promise<User | undefined> {
    return Array.from(this.users.values()).find(user => user.username === username);
  }

  async createUser(insertUser: InsertUser): Promise<User> {
    const id = this.currentUserId++;
    const user: User = { ...insertUser, id };
    this.users.set(id, user);
    return user;
  }

  // Menu Items
  async getAllMenuItems(): Promise<MenuItem[]> {
    return Array.from(this.menuItems.values());
  }

  async getMenuItemsByCategory(category: string): Promise<MenuItem[]> {
    return Array.from(this.menuItems.values()).filter(item => item.category === category);
  }

  async createMenuItem(insertItem: InsertMenuItem): Promise<MenuItem> {
    const id = this.currentMenuItemId++;
    const item: MenuItem = { ...insertItem, id };
    this.menuItems.set(id, item);
    return item;
  }

  // Orders
  async createOrder(insertOrder: InsertOrder): Promise<Order> {
    const id = this.currentOrderId++;
    const order: Order = { 
      ...insertOrder, 
      id, 
      createdAt: new Date()
    };
    this.orders.set(id, order);
    return order;
  }

  async getOrder(id: number): Promise<Order | undefined> {
    return this.orders.get(id);
  }

  async getAllOrders(): Promise<Order[]> {
    return Array.from(this.orders.values());
  }

  async updateOrderStatus(id: number, status: string): Promise<Order | undefined> {
    const order = this.orders.get(id);
    if (order) {
      order.status = status;
      this.orders.set(id, order);
      return order;
    }
    return undefined;
  }

  // Contact Messages
  async createContactMessage(insertMessage: InsertContactMessage): Promise<ContactMessage> {
    const id = this.currentContactMessageId++;
    const message: ContactMessage = { 
      ...insertMessage, 
      id, 
      createdAt: new Date()
    };
    this.contactMessages.set(id, message);
    return message;
  }

  async getAllContactMessages(): Promise<ContactMessage[]> {
    return Array.from(this.contactMessages.values());
  }
}

export const storage = new MemStorage();

File: server-go/storage.go
package main

import (
	"sync"
	"time"
)

type Storage struct {
	mu                      sync.RWMutex
	menuItems               map[int]MenuItem
	orders                  map[int]Order
	contactMessages         map[int]ContactMessage
	currentMenuItemID       int
	currentOrderID          int
	currentContactMessageID int
}

func NewStorage() *Storage {
	s := &Storage{
		menuItems:               make(map[int]MenuItem),
		orders:                  make(map[int]Order),
		contactMessages:         make(map[int]ContactMessage),
		currentMenuItemID:       1,
		currentOrderID:          1,
		currentContactMessageID: 1,
	}
	s.initializeMenuItems()
	return s
}

func (s *Storage) initializeMenuItems() {
	menuData := []MenuItem{
		{Name: "Samosas (2 pieces)", Description: "Crispy pastries filled with spiced potatoes and peas", Price: "6.99", Category: "Starters", Available: true},
		{Name: "Pakoras", Description: "Mixed vegetable fritters with mint chutney", Price: "8.99", Category: "Starters", Available: true},
		{Name: "Aloo Tikki", Description: "Pan-fried potato patties with tamarind sauce", Price: "7.99", Category: "Starters", Available: true},
		{Name: "Chana Chaat", Description: "Spiced chickpea salad with yogurt and chutneys", Price: "9.99", Category: "Starters", Available: true},
		{Name: "Butter Chicken", Description: "Tender chicken in rich tomato and cream sauce", Price: "16.99", Category: "Main Courses", Available: true},
		{Name: "Dal Makhani", Description: "Slow-cooked black lentils with butter and cream", Price: "14.99", Category: "Main Courses", Available: true},
		{Name: "Palak Paneer", Description: "Fresh spinach curry with cottage cheese", Price: "15.99", Category: "Main Courses", Available: true},
		{Name: "Chicken Biryani", Description: "Aromatic basmati rice with spices and chicken", Price: "18.99", Category: "Main Courses", Available: true},
		{Name: "Vegetable Biryani", Description: "Aromatic basmati rice with spices and vegetables", Price: "16.99", Category: "Main Courses", Available: true},
		{Name: "Gulab Jamun (2 pieces)", Description: "Soft milk dumplings in rose-scented syrup", Price: "5.99", Category: "Desserts", Available: true},
		{Name: "Kheer", Description: "Traditional rice pudding with cardamom and nuts", Price: "6.99", Category: "Desserts", Available: true},
	}

	for _, item := range menuData {
		id := s.currentMenuItemID
		item.ID = id
		s.menuItems[id] = item
		s.currentMenuItemID++
	}
}

func (s *Storage) GetAllMenuItems() []MenuItem {
	s.mu.RLock()
	defer s.mu.RUnlock()
	items := make([]MenuItem, 0, len(s.menuItems))
	for _, v := range s.menuItems {
		items = append(items, v)
	}
	return items
}

func (s *Storage) GetMenuItemsByCategory(category string) []MenuItem {
	s.mu.RLock()
	defer s.mu.RUnlock()
	items := []MenuItem{}
	for _, v := range s.menuItems {
		if v.Category == category {
			items = append(items, v)
		}
	}
	return items
}

func (s *Storage) CreateOrder(req CreateOrderRequest) Order {
	s.mu.Lock()
	defer s.mu.Unlock()
	order := Order{
		ID:                  s.currentOrderID,
		CustomerName:        req.CustomerName,
		CustomerPhone:       req.CustomerPhone,
		CustomerEmail:       req.CustomerEmail,
		CustomerAddress:     req.CustomerAddress,
		DeliveryDate:        req.DeliveryDate,
		DeliveryTime:        req.DeliveryTime,
		Items:               req.Items,
		SpecialInstructions: req.SpecialInstructions,
		Subtotal:            req.Subtotal,
		Tax:                 req.Tax,
		DeliveryFee:         req.DeliveryFee,
		Total:               req.Total,
		Status:              "Pending",
		CreatedAt:           time.Now(),
	}
	s.orders[s.currentOrderID] = order
	s.currentOrderID++
	return order
}

func (s *Storage) GetAllOrders() []Order {
	s.mu.RLock()
	defer s.mu.RUnlock()
	orders := make([]Order, 0, len(s.orders))
	for _, v := range s.orders {
		orders = append(orders, v)
	}
	return orders
}

func (s *Storage) GetOrder(id int) (Order, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	order, exists := s.orders[id]
	return order, exists
}

func (s *Storage) UpdateOrderStatus(id int, status string) (Order, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	order, exists := s.orders[id]
	if !exists {
		return Order{}, false
	}
	order.Status = status
	s.orders[id] = order
	return order, true
}

func (s *Storage) CreateContactMessage(req CreateContactRequest) ContactMessage {
	s.mu.Lock()
	defer s.mu.Unlock()
	message := ContactMessage{
		ID:        s.currentContactMessageID,
		Name:      req.Name,
		Email:     req.Email,
		Subject:   req.Subject,
		Message:   req.Message,
		CreatedAt: time.Now(),
	}
	s.contactMessages[s.currentContactMessageID] = message
	s.currentContactMessageID++
	return message
}

func (s *Storage) GetAllContactMessages() []ContactMessage {
	s.mu.RLock()
	defer s.mu.RUnlock()
	messages := make([]ContactMessage, 0, len(s.contactMessages))
	for _, v := range s.contactMessages {
		messages = append(messages, v)
	}
	return messages
}
